### PR TITLE
add basic quicksort support

### PR DESCRIPTION
**Pull request by pierrec**
The algorithm may be improved but I just want to get the basics in place.

### CODE REVIEW

File: lib/std/hash/crc32.c3 [hunk 1]
```diff
@@ -8,17 +8,17 @@ struct Crc32
     uint result;
 }
 
-fn void Crc32.init(Crc32* this, uint seed = 0)
+fn void Crc32.init(&this, uint seed = 0)
```
Comment (lerno) [inline lib/std/hash/crc32.c3:11]: Can you change this to `&self`? I found `&self` was more visually distinct and this is what I've been using pervasively.

File: lib/std/sort/quicksort.c3 [hunk 2]
```diff
@@ -0,0 +1,66 @@
+module std::sort;
+
+/**
+ * @require is_searchable(list) "The list must be indexable and support .len or .len()"
+ **/
+macro quicksort(list, $Type)
```
Comment (lerno) [inline lib/std/sort/quicksort.c3:6]: Why is the $Type here?

File: lib/std/sort/quicksort.c3 [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+ **/
+macro quicksort_with(list, $Type, cmp)
+{
+    (($Type)(list)).sort(cmp);
+}
+
+module std::sort::quicksort<Type, Comparer>;
+import std::sort;
+
+def Quicksort = distinct Type[];
```
Comment (lerno) [inline lib/std/sort/quicksort.c3:23]: Do we need to make a type?

**Comment by lerno** [comment]
I am thinking I'd probably just prefer to do:
```c
quicksort(array);
```
Shouldn't that be possible?

**Comment by pierrec** [comment]
I couldnt find a way without it to keep typed comparers.
The partition function needs a type unless we only use cmp(void*, void*).
Or am I missing something?

**Comment by lerno** [comment]
It should be possible, but I can massage this into shape if you want.

**Comment by pierrec** [comment]
Go ahead! Happy to learn more stuff.

**Comment by lerno** [comment]
You can have a look at the current one. Although arguably it's a bit repetitive. And it reveals something of a weakness in the system. I'll discuss it later on discord.

**Pull request merged**